### PR TITLE
feat(monaco): add theme hook

### PR DESCRIPTION
**Pull request by aionate0812**
### Overview

Follow up to: [add language service and delta decorations hook](https://github.com/Codecademy/client-modules/pull/1913) 

Monaco has two default themes, this PR ports/adds our custom themes for all the languages we support. It also adds a custom hook to update the editor's theme.

In the monolith, the theme hook makes use of user settings coming from redux to toggle high contrast. Because we want there are instances where we won't have user-specific settings I added a set of default settings that the hook can use. 

### PR Checklist

- [ ] Related to designs:
- [x] Related to JIRA ticket: [Reach-945](https://codecademy.atlassian.net/browse/REACH-945)
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.0.2-alpha.765e13.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60fb0b9ad6a9dc09fd578bcb--gamut-preview.netlify.app>https://60fb0b9ad6a9dc09fd578bcb--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60fb0b9ad6a9dc09fd578bcb>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.0.2-alpha.ff25a7.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60fb363124ae623d32140ec1--gamut-preview.netlify.app>https://60fb363124ae623d32140ec1--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60fb363124ae623d32140ec1>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.0.2-alpha.f71ab5.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6101854a02a53349f7c5a637--gamut-preview.netlify.app>https://6101854a02a53349f7c5a637--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6101854a02a53349f7c5a637>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.1.1-alpha.485725.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6101a52dfe672b1fb5701f68--gamut-preview.netlify.app>https://6101a52dfe672b1fb5701f68--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6101a52dfe672b1fb5701f68>Deploy Logs</a></p>

File: packages/monaco/src/MonacoEditor/index.tsx [hunk 1]
```diff
@@ -2,27 +2,45 @@ import React, { useState } from 'react';
...
+  screenReader: false,
 };
 
 export const MonacoEditor: React.FC<MonacoEditorProps> = ({
   file,
   onChange,
+  theme = MonacoThemeType.static,
+  userInterfaceSettings,
 }) => {
   // TODO import editor theming and settings hooks
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/MonacoEditor/index.tsx:32]:
```suggestion
  // TODO import editor settings hook
  
```

File: packages/monaco/src/MonacoEditor/useEditorTheming/index.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import { EditorTheme } from '../../libs/theme/editorTheme';
+import { Editor, UserInterfaceSettings } from '../types';
+
+export enum MonacoThemeType {
+  // 'static' here refers to the legacy LE monaco theme, which is a set mapping
+  // of colors, whereas 'dynamic' refers to a mapping that will change based on
+  // the learner's settings. E.g., dark mode vs light mode.
+  static = 'static',
+  dynamic = 'dynamic',
+}
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/MonacoEditor/useEditorTheming/index.ts:14]: This is actually new to me. @aionate0812 @casey-speer why not always go with the 'dynamic' theme, and have it be part of learner settings? Why would we want to explicitly override?

I'm not positive this should be ported to the shared module. I'd think the module want not want to know of such a difference, and instead be told its theme by the parent app.

File: packages/monaco/src/MonacoEditor/useEditorTheming/index.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+export const useEditorTheming = (
+  interfaceSettings: UserInterfaceSettings,
+  editor?: Editor.IStandaloneCodeEditor,
+  monaco?: Monaco,
+  themeType: MonacoThemeType = MonacoThemeType.static
+) => {
+  // TODO get user interface settings
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/MonacoEditor/useEditorTheming/index.ts:22]: ?

File: packages/monaco/src/libs/services/styles.module.scss [hunk 4]
```diff
@@ -0,0 +1,19 @@
+@import "~styles/cc";
+
+.lintDecoration {
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/libs/services/styles.module.scss:3]: bbllllurrrgghhhh this whole thing.

@dreamwasp @codecaaron is there a way to generate a class name in our Emotion / gamut-styles land? I forget...

File: packages/monaco/src/libs/theme/colorsLEThemed.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import { platformTheme as theme } from '@codecademy/gamut-styles';
+
+import { MonacoThemeColors } from './types';
+
+// TODO: get from learner settings when available
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/libs/theme/colorsLEThemed.ts:5]: ?

```suggestion
```

File: packages/monaco/src/libs/theme/colorsLEThemed.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+const colorMode = 'dark';
+const { _getColorValue: getColorValue } = theme;
+
+export const LEColors: MonacoThemeColors = {
+  syntax: {
+    attribute: getColorValue(theme.modes[colorMode]['editor-attribute']),
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/libs/theme/colorsLEThemed.ts:11]: Nit: I feel like you could deduplicate these to be a little less verbose...

```ts
const getColorValue = (color: string) => theme.getColorValue(theme.modes[colorMode][color]);
```

**Review by JoshuaKGoldberg — COMMENTED** [review]
I am surprised by there being a difference between LE and colors?

But since this is how it is in the monolith I think the only blocking-ish question for me is about the class names.

File: packages/monaco/src/MonacoEditor/useEditorTheming/index.ts (hunk 3 above)
Comment (aionate0812) [reply packages/monaco/src/MonacoEditor/useEditorTheming/index.ts]: not relevant anymore, removed

File: packages/monaco/src/libs/theme/colorsLEThemed.ts (hunk 6 above)
Comment (aionate0812) [reply packages/monaco/src/libs/theme/colorsLEThemed.ts]: typescript was not happy

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.1.1-alpha.f79768.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6102fbd961c2715ebe5013f6--gamut-preview.netlify.app>https://6102fbd961c2715ebe5013f6--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6102fbd961c2715ebe5013f6>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.2.1-alpha.1492b6.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://61113d284f319f8af6fcee6d--gamut-preview.netlify.app>https://61113d284f319f8af6fcee6d--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/61113d284f319f8af6fcee6d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.2.1-alpha.7d826e.0<br/>@codecademy/styleguide@43.7.4-alpha.7d826e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611bb04d29a61020086dea2d--gamut-preview.netlify.app>https://611bb04d29a61020086dea2d--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611bb04d29a61020086dea2d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/monaco@0.2.1-alpha.31ca91.0<br/>@codecademy/styleguide@43.7.4-alpha.31ca91.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://611bf167380b270e0e3076fc--gamut-preview.netlify.app>https://611bf167380b270e0e3076fc--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/611bf167380b270e0e3076fc>Deploy Logs</a></p>

File: packages/monaco/src/MonacoEditor/useEditorTheming/index.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    setThemeDefined((themeDefined) => {
+      if (!themeDefined) {
+        monaco.editor.defineTheme(
+          EditorTheme.dark,
+          themeType === MonacoThemeType.static ? darkTheme : CodecademyTheme
+        );
+      }
+
+      monaco.editor.setTheme(EditorTheme.dark);
```
Comment (judahtanthony) [inline packages/monaco/src/MonacoEditor/useEditorTheming/index.ts:43]: Can you help me understand this?  My assumptions:
1. defineTheme creates a name to config mapping.
2. setTheme activates a theme based on the previously defined name.
3. CodecademyTheme is one of darkTheme or lightTheme based on user context.
If that is the case, why aren't we just defining both light and dark, and then doing the `themeType===MonacoThemeType.static ...` when we are setting/choosing the theme?  If my assumptions are true, you might actually be giving the lightTheme the name EditorTheme.dark, just so we can show it.

Comment (judahtanthony) [reply packages/monaco/src/MonacoEditor/useEditorTheming/index.ts:43]: hmmm.  looking at packages/monaco/src/libs/theme/editorTheme.ts, it looks like my assumptions are off.  There is not lightTheme.

File: packages/styleguide/stories/Molecules/MonacoEditor.stories.mdx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    onChange: () => {},
+    readOnly: true,
+  }}
+/>
+
+<Canvas>
+  <Story name="MonacoEditor">
+    {(args) => (
+      <Box height="350px" width="100%">
+        <MonacoEditor {...args} />
```
Comment (judahtanthony) [inline packages/styleguide/stories/Molecules/MonacoEditor.stories.mdx:26]: Woohoo.  I'm so glad you were able to get this in storybook.

**Review by judahtanthony — COMMENTED** [review]
There is something slightly strange going on with this element on the right side, but I don't know if it is actually a bug.
Screen Shot 2021-08-20 at 11 53 32 AM

Also the readOnly prop in storybook doesn't seem to actually prevent you from altering the contents of the editor.  Am I mistaking the meaning of this?

File: packages/monaco/src/MonacoEditor/index.tsx [hunk 9]
```diff
@@ -2,27 +2,45 @@ import React, { useState } from 'react';
...
 export type MonacoEditorProps = {
   className?: string;
   file: MonacoFile;
   onChange?: (value: string) => void;
   readOnly?: boolean;
+  theme?: MonacoThemeType;
+  editorInterfaceSettings?: EditorInterfaceSettings;
+};
+
+export const defaultUserInterfaceSettings: EditorInterfaceSettings = {
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/MonacoEditor/index.tsx:18]: Nit:

```suggestion
const defaultUserInterfaceSettings: EditorInterfaceSettings = {
```

File: packages/monaco/src/MonacoEditor/types.ts [hunk 10]
```diff
@@ -5,3 +5,13 @@ export type MonacoFile = {
...
+
+export type EditorInterfaceSettings = {
+  highContrast: boolean;
+  screenReader: boolean;
+  autoCloseTokens: boolean;
+  renderWhitespace: boolean;
+  editorFontSize: EditorFontSize;
+};
+
+export type EditorFontSize = 'reg' | 'lg' | 'xl';
```
Comment (JoshuaKGoldberg) [inline packages/monaco/src/MonacoEditor/types.ts:17]: Can this be taken in from the Monaco types? Seems like it's one of theirs, no?

And why would we allow this? Are all of our editors not using the same font size?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
LGTM in general, thanks for the great coverage & docs!

Just waiting on the pending threads 👍

File: packages/monaco/src/MonacoEditor/useEditorTheming/index.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+      return true;
+    });
+  }, [highContrast, monaco, themeType]);
+
+  // We're overriding the monaco native toggleHighContrast action here.
+  // The original toggleHighContrastAction can be obtained by
+  // using editor.getAction('editor.action.toggleHighContrast').
+  // TODO add useEffect to update high contrast
```
Comment (casey-speer) [inline packages/monaco/src/MonacoEditor/useEditorTheming/index.ts:51]: What's up with this?

File: packages/monaco/src/libs/theme/colorsLEThemed.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+import { platformTheme as theme } from '@codecademy/gamut-styles';
```
Comment (casey-speer) [inline packages/monaco/src/libs/theme/colorsLEThemed.ts]: Similar to the points raised above, this is pretty monolith-specific at present. I guess this is kind of a philosophical point for now, but are `platformXXX` things intended as anything connected to the LE (as opposed to monolith-only)? Do we intend to have the code editor connected to the LE whenever we use it, or are there cases where we might want to use it standalone?  If the answer to either of those questions is "no", then we might not want to couple to platform-specific styles in this package.

**Comment by JoshuaKGoldberg** [comment]
Closing as aged away. I'll be continuing this push in January. Thanks!

**Pull request closed without merging**